### PR TITLE
fix AdaptiveAveragePooling crash problem for non support input

### LINKED ISSUE #51352: torch.nn.functional.adaptive_avg_pool3d and torch.nn.AdaptiveAvgPool3d aborts when input is of integral type
## 🐛 Bug

`torch.nn.functional.adaptive_avg_pool3d`  and `torch.nn.AdaptiveAvgPool3d` aborts when `input` contains negative number

## To Reproduce
~~~python
import torch
input = torch.tensor([[[[[-1]]]]])
torch.nn.functional.adaptive_avg_pool3d(input=input, output_size=1)
~~~

~~~python
import torch
torch.nn.AdaptiveAvgPool3d(1)( torch.tensor([[[[[-1]]]]]))
~~~

Output:
~~~python
terminate called without an active exception
Aborted (core dumped)
~~~

## Expected behavior

expect no crash

## Environment

PyTorch version: 1.8.0.dev20210126+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @mruberry

### PR DESCRIPTION
**Pull request by XiaobingSuper**
For none support input, we should not do check in a parallel region, this PR will first do the dtype check, and then do parallel for.
Fixes https://github.com/pytorch/pytorch/issues/51352.

### CODE REVIEW

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 1]
```diff
@@ -30,52 +30,51 @@ static void adaptive_avg_pool3d_out_frame(
...
-          /* compute local average: */
-          scalar_t sum = 0;
-          int it, ih, iw;
-          for (it = 0; it < kT; it++) {
-            for (ih = 0; ih < kH; ih++) {
-              for (iw = 0; iw < kW; iw++) {
-                scalar_t val =
-                    *(ip + it * istrideT + ih * istrideH + iw * istrideW);
-                sum += val;
+  at::parallel_for(0, sizeD, 1, [&](int64_t start, int64_t end) {
```
Comment (ngimel) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:33]: this parallel region is nested inside another one, I suspect it won't actually be parallelized, can you please check?
Good fix otherwise, we should get rid of #pragma omps. Can you please check if there are any other #pragma omps and get rid of them too?

Comment (XiaobingSuper) [reply aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:33]: Yes,  it is nested for dim=5, but for dim=4, it will be parallelized. I will check if there are any other #pragma omps.

Comment (XiaobingSuper) [reply aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:33]: https://github.com/pytorch/pytorch/blob/508bab43e71154b894cae04a6fa88547fd3c1505/aten/src/ATen/native/MaxUnpooling.cpp#L42 use #pragma omps, but there has a **#omp critical**, which needs refact the code to remove it(may do big change).  so merge this PR at first.

**Comment by ngimel** [comment]
Can you please add a test that would catch an error for unsupported type?

**Comment by XiaobingSuper** [comment]
> Can you please add a test that would catch an error for unsupported type?

Test case is added.

**Comment by XiaobingSuper** [comment]
@ngimel .the failed cases seem not relate to this PR.

**Review by ngimel — APPROVED** [review]
